File: app/src/main/java/com/douyin/downloaderqh/download/DownloadManager.kt
package com.douyin.downloaderqh.download

import android.content.ContentValues
import android.content.Context
import android.net.Uri
import android.os.Build
import android.provider.MediaStore
import com.douyin.downloaderqh.model.DownloadItem
import com.douyin.downloaderqh.model.DownloadType
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.withContext
import okhttp3.OkHttpClient
import okhttp3.Request
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.io.InputStream
import java.util.concurrent.TimeUnit

class DownloadManager(private val context: Context) {

    private val client = OkHttpClient.Builder()
        .connectTimeout(30, TimeUnit.SECONDS)
        .readTimeout(300, TimeUnit.SECONDS)
        .followRedirects(true)
        .build()

    /**
     * 完全在 IO 线程流式下载，每 500ms emit (进度 0..1, 速度字符串)。
     * 完成后拷贝到系统相册，最后 emit null 表示结束。
     */
    fun downloadWithProgress(
        item: DownloadItem,
        savePath: String?
    ): Flow<Pair<Float, String>?> = flow {
        val extension = when (item.type) {
            DownloadType.VIDEO, DownloadType.LIVE_PHOTO -> ".mp4"
            DownloadType.IMAGE -> ".jpg"
        }
        val safeTitle = item.title.replace(Regex("[/\\\\:*?\"<>|]"), "_").take(80)
        val fileName = "${safeTitle}$extension"

        val tempDir = File(context.cacheDir, "downloads")
        if (!tempDir.exists()) tempDir.mkdirs()
        val tempFile = File(tempDir, fileName)

        // ========== 所有网络和文件操作在 Dispatchers.IO 执行 ==========
        withContext(Dispatchers.IO) {
            val request = Request.Builder().url(item.url).build()
            val response = client.newCall(request).execute()

            if (!response.isSuccessful) {
                throw RuntimeException("下载失败 HTTP ${response.code}")
            }

            val body = response.body ?: throw RuntimeException("空响应体")
            val contentLength = body.contentLength()
            val inputStream: InputStream = body.byteStream()

            FileOutputStream(tempFile).use { fos ->
                val buffer = ByteArray(8192)
                var totalBytesRead = 0L
                var lastBytes = 0L
                var lastTime = System.currentTimeMillis()

                while (true) {
                    if (Thread.currentThread().isInterrupted) throw CancellationException()
                    val read = inputStream.read(buffer)
                    if (read == -1) break
                    fos.write(buffer, 0, read)
                    totalBytesRead += read

                    val now = System.currentTimeMillis()
                    val elapsed = now - lastTime
                    if (elapsed >= 500) {
                        // ---------- 进度 emit（可在主线程安全更新 UI）----------
                        val fraction = if (contentLength > 0) {
                            (totalBytesRead.toFloat() / contentLength).coerceIn(0f, 1f)
                        } else -1f
                        val speedStr = if (elapsed > 0) {
                            val bytesDelta = totalBytesRead - lastBytes
                            val bytesPerSecond = (bytesDelta * 1000L / elapsed)
                            formatSpeed(bytesPerSecond)
                        } else "—"
                        lastBytes = totalBytesRead
                        lastTime = now
                        emit(fraction to speedStr)
                    }
                }
            }

            response.close()

            try {
                moveToGallery(tempFile, fileName, item.type, savePath)
            } catch (_: Exception) {
                // 拷贝失败不抛异常，文件至少已在缓存目录
            }

            emit(null)   // 完成信号
        }
    }

    private fun moveToGallery(
        file: File,
        fileName: String,
        type: DownloadType,
        savePath: String?
    ): String? {
        if (!file.exists() || file.length() == 0L) return null

        val mime = when (type) {
            DownloadType.VIDEO, DownloadType.LIVE_PHOTO -> "video/mp4"
            DownloadType.IMAGE -> "image/jpeg"
        }
        val cv = ContentValues().apply {
            put(MediaStore.MediaColumns.DISPLAY_NAME, fileName)
            put(MediaStore.MediaColumns.MIME_TYPE, mime)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
                put(MediaStore.MediaColumns.RELATIVE_PATH,
                    if (type == DownloadType.IMAGE) "Pictures/Douyin" else "Movies/Douyin")
            }
        }
        val collection = if (type == DownloadType.IMAGE) MediaStore.Images.Media.EXTERNAL_CONTENT_URI
        else MediaStore.Video.Media.EXTERNAL_CONTENT_URI

        val uri = context.contentResolver.insert(collection, cv) ?: return null
        context.contentResolver.openOutputStream(uri)?.use { out ->
            FileInputStream(file).use { fis ->
                fis.copyTo(out, 8192)
            }
        }
        file.delete()
        return uri.toString()
    }

    private fun formatSpeed(bytesPerSecond: Long): String = when {
        bytesPerSecond < 1024 -> "$bytesPerSecond B/s"
        bytesPerSecond < 1024 * 1024 -> "${"%.1f".format(bytesPerSecond / 1024.0)} KB/s"
        bytesPerSecond < 1024 * 1024 * 1024 -> "${"%.1f".format(bytesPerSecond / (1024.0 * 1024))} MB/s"
        else -> "${"%.2f".format(bytesPerSecond / (1024.0 * 1024 * 1024))} GB/s"
    }
}

File: lib/download/download_manager.dart
import 'dart:async';
import 'dart:io';
import 'package:http/http.dart' as http;
import 'package:path_provider/path_provider.dart';
import '../model/models.dart';
import 'package:image_gallery_saver_plus/image_gallery_saver_plus.dart';

class DownloadManager {
  final http.Client _client;

  DownloadManager({http.Client? client}) : _client = client ?? http.Client();

  Stream<DownloadProgress> downloadItem(DownloadItem item) async* {
    final extension = switch (item.type) {
      DownloadType.video || DownloadType.livePhoto => '.mp4',
      DownloadType.image => '.jpg',
    };
    final safeTitle = item.title.replaceAll(RegExp(r'[/\\:*?"<>|]'), '_');
    final fileName = '${safeTitle.length > 80 ? safeTitle.substring(0, 80) : safeTitle}$extension';

    // 获取临时目录
    final tempDir = await getTemporaryDirectory();
    final tempFile = File('${tempDir.path}/downloads/$fileName');
    await tempFile.parent.create(recursive: true);

    try {
      final request = http.Request('GET', Uri.parse(item.url));
      final streamedResponse = await _client.send(request);

      if (streamedResponse.statusCode != 200) {
        throw Exception('下载失败 HTTP ${streamedResponse.statusCode}');
      }

      final contentLength = streamedResponse.contentLength ?? -1;
      final sink = tempFile.openWrite();
      int totalBytesRead = 0;
      int lastBytes = 0;
      int lastTime = DateTime.now().millisecondsSinceEpoch;

      await for (final chunk in streamedResponse.stream) {
        sink.add(chunk);
        totalBytesRead += chunk.length;

        final now = DateTime.now().millisecondsSinceEpoch;
        final elapsed = now - lastTime;

        if (elapsed >= 500) {
          final fraction = contentLength > 0
              ? (totalBytesRead / contentLength).clamp(0.0, 1.0)
              : -1.0;
          final speedStr = elapsed > 0
              ? _formatSpeed((totalBytesRead - lastBytes) * 1000 ~/ elapsed)
              : '—';
          lastBytes = totalBytesRead;
          lastTime = now;

          yield DownloadProgress(
            fraction: fraction,
            speed: speedStr,
            bytesDownloaded: totalBytesRead,
            totalBytes: contentLength,
          );
        }
      }

      await sink.flush();
      await sink.close();

      // 保存到相册
      final result = await _saveToGallery(tempFile, fileName, item.type);

      // 删除临时文件
      await tempFile.delete();

      yield DownloadProgress(
        fraction: 1.0,
        speed: '',
        bytesDownloaded: totalBytesRead,
        totalBytes: contentLength,
        completed: true,
        savedPath: result,
      );
    } catch (e) {
      // 清理临时文件
      if (await tempFile.exists()) {
        await tempFile.delete();
      }
      yield DownloadProgress(
        fraction: 0,
        speed: '',
        bytesDownloaded: 0,
        totalBytes: 0,
        completed: true,
        error: e.toString(),
      );
    }
  }

  Future<String> _saveToGallery(File file, String fileName, DownloadType type) async {
    final mimeType = switch (type) {
      DownloadType.video || DownloadType.livePhoto => 'video/mp4',
      DownloadType.image => 'image/jpeg',
    };

    final result = await ImageGallerySaverPlus.saveFile(
      file.path,
      name: fileName,
      isReturnPathOfIOS: true,
    );

    if (result == null) {
      throw Exception('保存到相册失败');
    }

    return result.toString();
  }

  String _formatSpeed(int bytesPerSecond) {
    if (bytesPerSecond < 1024) return '$bytesPerSecond B/s';
    if (bytesPerSecond < 1024 * 1024) {
      return '${(bytesPerSecond / 1024).toStringAsFixed(1)} KB/s';
    }
    if (bytesPerSecond < 1024 * 1024 * 1024) {
      return '${(bytesPerSecond / (1024 * 1024)).toStringAsFixed(1)} MB/s';
    }
    return '${(bytesPerSecond / (1024 * 1024 * 1024)).toStringAsFixed(2)} GB/s';
  }

  void dispose() {
    _client.close();
  }
}

class DownloadProgress {
  final double fraction;
  final String speed;
  final int bytesDownloaded;
  final int totalBytes;
  final bool completed;
  final String? savedPath;
  final String? error;

  DownloadProgress({
    required this.fraction,
    required this.speed,
    required this.bytesDownloaded,
    required this.totalBytes,
    this.completed = false,
    this.savedPath,
    this.error,
  });

  bool get isError => error != null;
  bool get isDone => completed && error == null;
}

extension _StringTake on String {
  String take(int maxLength) {
    if (length <= maxLength) return this;
    return substring(0, maxLength);
  }
}
